File: code/java/ch15/learn/Example02SlidingWindowPatterns.java
package ch15.learn;

import java.util.*;

/**
 * Example 02: Sliding Window Patterns
 * =====================================
 * Chapter 15: Two Pointers & Sliding Window — The Caterpillar Method
 *
 * Demonstrates fixed-size and variable-size sliding windows.
 */
public class Example02SlidingWindowPatterns {

    public static void main(String[] args) {
        // Part 1: Fixed-size window — max sum of k elements
        System.out.println("=== Part 1: Fixed-Size Window ===");
        int[] arr = {2, 1, 5, 1, 3, 2, 8, 1};
        int k = 3;
        System.out.println("Array: [2,1,5,1,3,2,8,1], k=" + k);

        int windowSum = 0;
        for (int i = 0; i < k; i++) windowSum += arr[i];
        int best = windowSum;

        for (int i = k; i < arr.length; i++) {
            windowSum += arr[i] - arr[i - k];
            best = Math.max(best, windowSum);
            System.out.printf("  Window [%d..%d] sum=%d%s%n",
                i - k + 1, i, windowSum, windowSum == best ? " <- best" : "");
        }
        System.out.println("Best sum: " + best);

        // Part 2: Variable-size window — longest substring without repeats
        System.out.println("\n=== Part 2: Variable Window + HashMap ===");
        String s = "abcabcbb";
        System.out.println("String: \"" + s + "\"");

        Map<Character, Integer> charIndex = new HashMap<>();
        int left = 0, bestLen = 0;

        for (int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);
            if (charIndex.containsKey(ch) && charIndex.get(ch) >= left) {
                left = charIndex.get(ch) + 1;
            }
            charIndex.put(ch, right);
            bestLen = Math.max(bestLen, right - left + 1);
            System.out.printf("  right=%d '%c' window=\"%s\" len=%d%n",
                right, ch, s.substring(left, right + 1), right - left + 1);
        }
        System.out.println("Longest substring without repeating: " + bestLen);
    }
}

File: code/cpp/ch15/learn/example_02_sliding_window_patterns.cpp
/*
 * Example 02: Sliding Window Patterns
 * =====================================
 * Chapter 15: Two Pointers & Sliding Window — The Caterpillar Method
 *
 * Demonstrates fixed-size and variable-size sliding windows.
 */

#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

int main() {
    // Part 1: Fixed-size window — max sum of k elements
    cout << "=== Part 1: Fixed-Size Window ===" << endl;
    vector<int> arr = {2, 1, 5, 1, 3, 2, 8, 1};
    int k = 3;

    int windowSum = 0;
    for (int i = 0; i < k; i++) windowSum += arr[i];
    int best = windowSum;

    for (int i = k; i < (int)arr.size(); i++) {
        windowSum += arr[i] - arr[i - k];
        best = max(best, windowSum);
        cout << "  Window [" << i - k + 1 << ".." << i << "] sum=" << windowSum
             << (windowSum == best ? " <- best" : "") << endl;
    }
    cout << "Best sum: " << best << endl;

    // Part 2: Variable window — longest substring without repeats
    cout << "\n=== Part 2: Variable Window + HashMap ===" << endl;
    string s = "abcabcbb";
    cout << "String: \"" << s << "\"" << endl;

    unordered_map<char, int> charIndex;
    int left = 0, bestLen = 0;

    for (int right = 0; right < (int)s.size(); right++) {
        char ch = s[right];
        if (charIndex.count(ch) && charIndex[ch] >= left) {
            left = charIndex[ch] + 1;
        }
        charIndex[ch] = right;
        bestLen = max(bestLen, right - left + 1);
        cout << "  right=" << right << " '" << ch
             << "' window=\"" << s.substr(left, right - left + 1)
             << "\" len=" << right - left + 1 << endl;
    }
    cout << "Longest substring without repeating: " << bestLen << endl;

    return 0;
}
